### PR TITLE
Support emojis and reactions

### LINKED ISSUE #21: Fix emojis and reactions
Right now all code that supports emoji lookups and reactions is disabled. This issue will resolve the difficulties that caused them to be temporarily removed, as well as add new code for handling those events. (e.g. adding to `reactions` when crawling messages)

One of the issues are IDs. There are two kinds of emojis: "normal" unicode ones, which are just unicode code points, and custom emojis. While the first are simply characters, the second are Discord objects with a snowflake, associated guild, etc.

Previously I just made the ID of a unicode emoji it's `ord()` value. But this is problematic if you want to `get_emoji(id)` or perform other operations on the ID.

To resolve this, I propose making the emoji ID effectively an algebraic data type. It will have a primary key composed of two columns, a unicode character, and an ID. Exactly one of the two will be null, and the one with the value will show which type it is. For a unicode emoji, it's "ID" will be the unicode character itself.

**Schema proposals:**
`emojis`:

| `emoji_id`             | `emoji_unicode`        | `name`   | `category` | `guild_id`           |
|------------------------|------------------------|----------|------------|----------------------|
| `Optional[BigInteger]` | `Optional[Unicode(1)]` | `String` | `String`   | `Optional[BigInteger]` |

`reactions`:

| `timestamp` | `message_id` | `emoji_id`             | `emoji_unicode`        | `user_id`    | `channel_id` | `guild_id`   |
|-------------|--------------|------------------------|------------------------|--------------|--------------|--------------|
| `DateTime`  | `BigInteger` | `Optional[BigInteger]` | `Optional[Unicode(1)]` | `BigInteger` | `BigInteger` | `BigInteger` |

### PR DESCRIPTION
**Pull request by ammongit**
Fixes #21 

Discord emojis are quite a pain, but this PR finally adds reactions and emojis and tracks them in two tables: `emojis` and `reactions`.
Due to the nature of discord emojis either being either images with an attached snowflake, unicode characters, or even unicode strings, the "primary key" used to refer to one is actually two columns: `emoji_id` and `emoji_unicode`. Neither is nullable due to weirdness involving constraints on null columns, so instead if an emoji is custom, the `emoji_unicode` field is an empty string, and if the emoji is unicode-based, the `emoji_id` field is 0.

Also added internally is the `EmojiData` class, which helps isolate all the emoji managing weirdness to a single structure.

This tracks reactions and emojis both as live events and while traversing the history. However, reactions retrieved from looking at the history have less information than live ones, and any fields where data is unknown is set to null.

### CODE REVIEW

File: statbot/sql.py [hunk 1]
```diff
@@ -915,56 +911,53 @@ def upsert_member(self, trans, member):
...
-        # pylint: disable=unreachable
-        raise NotImplementedError
-
-        values = emoji_values(emoji)
-        id = values['emoji_id']
-        if id in self.emoji_cache:
-            self.logger.debug(f"Emoji {id} already inserted.")
+        data = EmojiData(emoji)
+        if data.cache_id in self.emoji_cache:
+            self.logger.debug(f"Emoji {data} already inserted.")
```
Comment (raylu) [inline statbot/sql.py]: can we talk about how this plus `_Transaction` just re-invents sqlalchemy sessions?

Comment (ammongit) [reply statbot/sql.py]: Yes, but according to the SQLAlchemy docs, sessions should be a more permanent thing, removed from individual clusters of operations. Given that we need this handle to be created and closed for every single operation statbot does, using transactions is more suitable.

**Comment by raylu** [comment]
this seems to track emoji only as reactions
https://discordapp.com/developers/docs/resources/channel#message-object has a `reactions?` bit which links to https://discordapp.com/developers/docs/resources/emoji#emoji-object
which implies that every emoji reaction will have a snowflake id. is this not the case? (do we need this crazy unicode dual key thing?)

File: statbot/emoji.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            self.category = ['custom']
+            self.roles = getattr(emoji, 'roles', None)
+            self.guild = getattr(emoji, 'guild', None)
+
+    @property
+    def mention(self):
+        if self.id:
+            return f'<:{self.name[0]}:{self.id}>'
+        else:
+            return self.unicode
```
Comment (raylu) [inline statbot/emoji.py]: what is this function?

**Comment by ammongit** [comment]
@raylu Unfortunately emojis are more complicated than that. The IDs apply only for emojis that are custom, like :strinking:. But for unicode emojis, those are passed as unicode strings. And I say "strings" instead of "characters" since not all emojis are a single character. For instance the US flag is two characters, and :family_mmbb: is actually _seven_ characters long.

That's why I put off adding this support for so long: it's a big pain.

File: statbot/emoji.py (hunk 2 above)
Comment (ammongit) [reply statbot/emoji.py]: It produces the "mention" for an emoji. Basically its raw representation in a string.

Comment (raylu) [reply statbot/emoji.py]: is it used anywhere?

Comment (ammongit) [reply statbot/emoji.py]: No. It's sort of just meant as "completion" to what is basically the emoji class.

Comment (raylu) [reply statbot/emoji.py]: I don't really understand why this is here. what's the downside to removing it?

Comment (ammongit) [reply statbot/emoji.py]: None right now, but it may be useful. Though it will probably end up being moved the client library.

**Pull request merged**